File: packages/node/iii/src/channels.ts
import { Readable, Writable } from 'node:stream'
import { WebSocket } from 'ws'
import type { StreamChannelRef } from './iii-types'

export class ChannelWriter {
  private static readonly FRAME_SIZE = 64 * 1024
  private ws: WebSocket | null = null
  private wsReady = false
  private readonly pendingMessages: {
    data: Buffer | string
    callback: (err?: Error | null) => void
  }[] = []
  public readonly stream: Writable
  private readonly url: string

  constructor(engineWsBase: string, ref: StreamChannelRef) {
    this.url = buildChannelUrl(engineWsBase, ref.channel_id, ref.access_key, 'write')

    this.stream = new Writable({
      write: (chunk: Buffer, _encoding, callback) => {
        const buf = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk)
        this.sendChunked(buf, callback)
      },
      final: callback => {
        if (!this.ws) {
          callback()
          return
        }
        if (this.wsReady) {
          this.ws.close(1000, 'stream_complete')
        } else {
          this.ws.on('open', () => this.ws?.close(1000, 'stream_complete'))
        }
        callback()
      },
      destroy: (err, callback) => {
        if (this.ws) this.ws.terminate()
        callback(err)
      },
    })
  }

  private ensureConnected(): void {
    if (this.ws) return
    this.ws = new WebSocket(this.url)

    this.ws.on('open', () => {
      this.wsReady = true
      for (const { data, callback } of this.pendingMessages) {
        this.ws?.send(data, callback)
      }
      this.pendingMessages.length = 0
    })

    this.ws.on('error', err => {
      this.stream.destroy(err)
    })

    this.ws.on('close', () => {
      if (!this.stream.destroyed) {
        this.stream.destroy()
      }
    })
  }

  sendMessage(msg: string): void {
    this.ensureConnected()
    this.sendRaw(msg, err => {
      if (err) this.stream.destroy(err)
    })
  }

  close(): void {
    if (!this.ws) return
    if (this.wsReady) {
      this.ws.close(1000, 'channel_close')
    } else {
      this.ws.on('open', () => this.ws?.close(1000, 'channel_close'))
    }
  }

  private sendChunked(data: Buffer, callback: (err?: Error | null) => void): void {
    let offset = 0
    const sendNext = (err?: Error | null): void => {
      if (err) {
        callback(err)
        return
      }

      if (offset >= data.length) {
        callback(null)
        return
      }

      const end = Math.min(offset + ChannelWriter.FRAME_SIZE, data.length)
      const part = data.subarray(offset, end)
      offset = end
      this.sendRaw(part, sendNext)
    }
    sendNext(null)
  }

  private sendRaw(data: Buffer | string, callback: (err?: Error | null) => void): void {
    this.ensureConnected()
    if (this.wsReady && this.ws) {
      this.ws.send(data, err => callback(err ?? null))
    } else {
      this.pendingMessages.push({ data, callback })
    }
  }
}

export class ChannelReader {
  private ws: WebSocket | null = null
  private connected = false
  private readonly messageCallbacks: Array<(msg: string) => void> = []
  public readonly stream: Readable
  private readonly url: string

  constructor(engineWsBase: string, ref: StreamChannelRef) {
    this.url = buildChannelUrl(engineWsBase, ref.channel_id, ref.access_key, 'read')

    const self = this
    this.stream = new Readable({
      read() {
        self.ensureConnected()
        if (self.ws) self.ws.resume()
      },
      destroy(err, callback) {
        if (self.ws && self.ws.readyState !== WebSocket.CLOSED) {
          self.ws.terminate()
        }
        self.ws = null
        callback(err)
      },
    })
  }

  private ensureConnected(): void {
    if (this.connected) return
    this.connected = true
    this.ws = new WebSocket(this.url)

    this.ws.on('open', () => {
      ;(this.ws as unknown as { binaryType: string }).binaryType = 'nodebuffer'
    })

    this.ws.on('message', (data: Buffer, isBinary: boolean) => {
      if (isBinary) {
        if (!this.stream.push(data)) {
          this.ws?.pause()
        }
      } else {
        const msg = data.toString('utf-8')
        for (const cb of this.messageCallbacks) {
          cb(msg)
        }
      }
    })

    this.ws.on('close', () => {
      this.ws = null
      if (!this.stream.destroyed) this.stream.push(null)
    })

    this.ws.on('error', err => {
      this.stream.destroy(err)
    })
  }

  onMessage(callback: (msg: string) => void): void {
    this.messageCallbacks.push(callback)
  }
}

function buildChannelUrl(
  engineWsBase: string,
  channelId: string,
  accessKey: string,
  direction: 'read' | 'write',
): string {
  const base = engineWsBase.replace(/\/$/, '')
  return `${base}/ws/channels/${channelId}?key=${encodeURIComponent(accessKey)}&dir=${direction}`
}

File: packages/rust/iii/src/channels.rs
use std::sync::Arc;

use futures_util::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tokio::sync::Mutex;
use tokio_tungstenite::{connect_async, tungstenite::Message as WsMessage};

use crate::error::IIIError;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum ChannelDirection {
    #[default]
    Read,
    Write,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct StreamChannelRef {
    pub channel_id: String,
    pub access_key: String,
    pub direction: ChannelDirection,
}

#[derive(Debug, Clone)]
pub enum ChannelItem {
    Text(String),
    Binary(Vec<u8>),
}

type WsWriter = futures_util::stream::SplitSink<
    tokio_tungstenite::WebSocketStream<tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>>,
    WsMessage,
>;

type WsReader = futures_util::stream::SplitStream<
    tokio_tungstenite::WebSocketStream<tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>>,
>;

type MessageCallback = Box<dyn Fn(String) + Send + Sync>;
type MessageCallbacks = Arc<Mutex<Vec<MessageCallback>>>;

fn build_channel_url(
    engine_ws_base: &str,
    channel_id: &str,
    access_key: &str,
    direction: &str,
) -> String {
    let base = engine_ws_base.trim_end_matches('/');
    let encoded_key = urlencoded(access_key);
    format!("{base}/ws/channels/{channel_id}?key={encoded_key}&dir={direction}")
}

fn urlencoded(s: &str) -> String {
    let mut result = String::with_capacity(s.len());
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                result.push(b as char);
            }
            _ => {
                result.push('%');
                result.push(char::from(b"0123456789ABCDEF"[(b >> 4) as usize]));
                result.push(char::from(b"0123456789ABCDEF"[(b & 0x0F) as usize]));
            }
        }
    }
    result
}

/// WebSocket-backed writer for streaming binary data and text messages.
pub struct ChannelWriter {
    url: String,
    ws: Arc<Mutex<Option<WsWriter>>>,
}

impl ChannelWriter {
    pub fn new(engine_ws_base: &str, channel_ref: &StreamChannelRef) -> Self {
        Self {
            url: build_channel_url(
                engine_ws_base,
                &channel_ref.channel_id,
                &channel_ref.access_key,
                "write",
            ),
            ws: Arc::new(Mutex::new(None)),
        }
    }

    async fn ensure_connected(&self) -> Result<(), IIIError> {
        let mut guard = self.ws.lock().await;
        if guard.is_some() {
            return Ok(());
        }
        let (stream, _) = connect_async(&self.url).await?;
        let (writer, _reader) = stream.split();
        *guard = Some(writer);
        Ok(())
    }

    const MAX_FRAME_SIZE: usize = 64 * 1024;

    pub async fn write(&self, data: &[u8]) -> Result<(), IIIError> {
        self.ensure_connected().await?;
        let mut guard = self.ws.lock().await;
        if let Some(ws) = guard.as_mut() {
            for chunk in data.chunks(Self::MAX_FRAME_SIZE) {
                ws.send(WsMessage::Binary(chunk.to_vec().into())).await?;
            }
        }
        Ok(())
    }

    pub async fn send_message(&self, msg: &str) -> Result<(), IIIError> {
        self.ensure_connected().await?;
        let mut guard = self.ws.lock().await;
        if let Some(ws) = guard.as_mut() {
            ws.send(WsMessage::Text(msg.to_string().into())).await?;
        }
        Ok(())
    }

    pub async fn close(&self) -> Result<(), IIIError> {
        let mut guard = self.ws.lock().await;
        if let Some(ws) = guard.as_mut() {
            ws.send(WsMessage::Close(None)).await?;
        }
        *guard = None;
        Ok(())
    }
}

/// WebSocket-backed reader for streaming binary data and text messages.
pub struct ChannelReader {
    url: String,
    ws: Arc<Mutex<Option<WsReader>>>,
    message_callbacks: MessageCallbacks,
}

impl ChannelReader {
    pub fn new(engine_ws_base: &str, channel_ref: &StreamChannelRef) -> Self {
        Self {
            url: build_channel_url(
                engine_ws_base,
                &channel_ref.channel_id,
                &channel_ref.access_key,
                "read",
            ),
            ws: Arc::new(Mutex::new(None)),
            message_callbacks: Arc::new(Mutex::new(Vec::new())),
        }
    }

    async fn ensure_connected(&self) -> Result<(), IIIError> {
        let mut guard = self.ws.lock().await;
        if guard.is_some() {
            return Ok(());
        }
        let (stream, _) = connect_async(&self.url).await?;
        let (_writer, reader) = stream.split();
        *guard = Some(reader);
        Ok(())
    }

    /// Register a callback for text messages received on this channel.
    pub async fn on_message<F>(&self, callback: F)
    where
        F: Fn(String) + Send + Sync + 'static,
    {
        self.message_callbacks.lock().await.push(Box::new(callback));
    }

    /// Read the next binary chunk from the channel.
    /// Text messages are dispatched to registered callbacks.
    /// Returns `None` when the stream is closed.
    pub async fn next_binary(&self) -> Result<Option<Vec<u8>>, IIIError> {
        self.ensure_connected().await?;
        let mut guard = self.ws.lock().await;
        let reader = guard.as_mut().ok_or(IIIError::NotConnected)?;

        loop {
            match reader.next().await {
                Some(Ok(WsMessage::Binary(data))) => return Ok(Some(data.to_vec())),
                Some(Ok(WsMessage::Text(text))) => {
                    let callbacks = self.message_callbacks.lock().await;
                    for cb in callbacks.iter() {
                        cb(text.to_string());
                    }
                }
                Some(Ok(WsMessage::Close(_))) | None => return Ok(None),
                Some(Ok(_)) => continue,
                Some(Err(e)) => return Err(IIIError::WebSocket(e.to_string())),
            }
        }
    }

    /// Read the entire stream into a single Vec<u8>.
    pub async fn read_all(&self) -> Result<Vec<u8>, IIIError> {
        let mut buffer = Vec::new();
        while let Some(chunk) = self.next_binary().await? {
            buffer.extend_from_slice(&chunk);
        }
        Ok(buffer)
    }

    pub async fn close(&self) -> Result<(), IIIError> {
        let mut guard = self.ws.lock().await;
        *guard = None;
        Ok(())
    }
}

/// Check if a JSON value looks like a StreamChannelRef.
pub fn is_channel_ref(value: &Value) -> bool {
    value.is_object()
        && value.get("channel_id").is_some_and(|v| v.is_string())
        && value.get("access_key").is_some_and(|v| v.is_string())
        && value.get("direction").is_some_and(|v| v.is_string())
}

/// Extract all channel references from a JSON value's top-level fields,
/// returning the field path and the deserialized ref.
pub fn extract_channel_refs(data: &Value) -> Vec<(String, StreamChannelRef)> {
    let mut refs = Vec::new();
    extract_refs_recursive(data, String::new(), &mut refs);
    refs
}

fn extract_refs_recursive(
    data: &Value,
    prefix: String,
    refs: &mut Vec<(String, StreamChannelRef)>,
) {
    if let Some(obj) = data.as_object() {
        for (key, value) in obj {
            let path = if prefix.is_empty() {
                key.clone()
            } else {
                format!("{prefix}.{key}")
            };

            if is_channel_ref(value) {
                if let Ok(channel_ref) = serde_json::from_value::<StreamChannelRef>(value.clone()) {
                    refs.push((path, channel_ref));
                }
            } else if value.is_object() {
                extract_refs_recursive(value, path.clone(), refs);
            } else if let Some(arr) = value.as_array() {
                for (idx, item) in arr.iter().enumerate() {
                    extract_refs_recursive(item, format!("{path}[{idx}]"), refs);
                }
            }
        }
    } else if let Some(arr) = data.as_array() {
        for (idx, item) in arr.iter().enumerate() {
            let path = if prefix.is_empty() {
                format!("[{idx}]")
            } else {
                format!("{prefix}[{idx}]")
            };
            extract_refs_recursive(item, path, refs);
        }
    }
}
